### ISSUE TITLE
Choosing of string to be used for plugin compatibility with 2.0

### ISSUE THREAD

**Issue by iionly** [Elgg/community_plugins#126]
Simpliest choice would be to select "2.0" as the string plugins compatible with Elgg 2.0.

The question is if the string "2.0" would be the best choice with semver versioning of Elgg core now in place or if it might be better to select something different for this release, for example just "2" or "2.X.X".

Any plugin that works on 2.0 should also work on any following 2.X release of Elgg (or there's something wrong in Elgg core, right?). But if the plugin compatibility on the community page says "2.0" it will result in the usual questions like "does this plugin work also on Elgg 2.1?" etc.

Using "2" or "2.X.X" could clarify this confusion to a certain extend. On the other hand, it could also increase the confusion as some people might not understand what "2.X.X" is supposed to mean. Also, I think the community_plugins plugin would currently not be able to handle the plugin compatibility sufficiently. For example, searching for plugins compatible with Elgg 2.1 would not return plugins that are compatible in principle but not explicitely marked as compatible with 2.1 but only with 2.0 or 2.X.X.

So, the questions: should the plugin compatibility with Elgg 2.0 be handled differently (at least regarding the string to be used) or should it just be handled like any other release of Elgg?

**Comment by ewinslow** [comment]
How about `2.0+`?

**Comment by ewinslow** [comment]
Then you'd have `2.1+`, `2.2+` as minimum versions increase

**Comment by ewinslow** [comment]
Alternatively, we could just use composer version identifiers directly (e.g., `^2.0`), with a link to "what does this mean?"

**Comment by iionly** [comment]
If a string is used that indicates something like "2.0 or newer" it might also be necessary to show that the "or newer" is limited to a newer 2.X version only and not necessarily any new Elgg version (3.X etc.). With "2.0+" alone this might not be clear.

I don't know if "2.0 or newer 2.X" would be a too long string. And I think it makes only sense to use any kind of string if it also works with regards to resolve the compatibiity with future versions correctly (e.g. in plugin search).

I don't like any references to composer (e.g. notation) used on the community site to indicate plugin compatibility. Downloading and using plugin zips from the community site has nothing to do with composer.

**Comment by hypeJunction** [comment]
My understanding is that 2.x will contain new features, which plugins maybe
using, so each version should have its own entry, despite semver. I don't
really understand what the discussion is about

**Comment by iionly** [comment]
@hypeJunction There might be new features introduced for example in 2.1 resulting in a plugin that makes use of these new features not working on 2.0. But that's not true the other way round, i.e. a plugin that works on 2.0 should still work on 2.1 without issues due to semver.

Currently, it happens again and again that people ask "does plugin x work on Elgg version Y?" and that's what this discussion is about: would there be any way to better show the users of a plugin that it also works on a future version of Elgg (specifically a new 2.X version as semver is now supported right from the start of this major Elgg version) without the plugin developers having to update the compatibility information of their plugins every few months again and again.

**Comment by hypeJunction** [comment]
Aha, I see now. Perhaps switching to range search input would make it clearer, e.g. Find plugins for Elgg versions between ___ and ___, and let the backend do the crunching.

**Comment by ewinslow** [comment]
I feel like we should just do what packagist does and pull the value straight out of composer.json

**Comment by ewinslow** [comment]
This exposed folks to tildes and carets, but it's something that is standard in the php world and we don't have to invent or maintain it.

There should also just be an easy way to ask "is it compatible with version x.y.z?"

**Comment by iionly** [comment]
How's ```^2.0``` any easier to understand than "2.0"? I would even say it's much worse.

It's not the very small number of developers who might have difficulties to understand to which versions of Elgg a plugin is compatible to but rather the larger number of non-developers who might not know this. If you are new to Elgg and have no or only limited experience with developing code, would it really feel good having to learn about composer, coding etc. if you have no intention to code on your own but only use what is ready it to used?

**Comment by ewinslow** [comment]
I didn't say it was easier? All I said was it's a de facto standard that we don't have to invent or maintain.

You don't have to learn composer or coding to learn that the caret means "everything from the given version up to (but not including) the next major version". We could provide a ui to check compatibility with a given version and return a yes/no answer. Then they don't have to think about the format at all.

Any symbol we choose is going to be arbitrary/ambiguous. At least if we go with composer's standard then the knowledge gaining by learning our conventions will be transferable to other projects too.

It's also one less thing for plugin devs or us to maintain since we can just pull the value straight out of the composer.json file instead of forcing people to input the value a second time in a different format which we deemed to be " easier for Non technical users " even though there is no way we have the resources to do the research needed to actually prove that it is any easier let alone so substantially easier that it merits burdening the developers in our ecosystem with extra busy work.

And last I checked, you were complaining that 2.0 is not clear enough to be an option.

Suggest an alternative and we can discuss it. If you don't have an alternative, then the default way forward will be composer style version constraints.

**Comment by ewinslow** [comment]
Please forgive the tone of the above comment. After reflecting I believe I got too aggressive/forceful.

Decisions like this are often a balance between usability for Non-devs and maintenance burden for devs. In this case I'm inclined to favor devs interests because it doesn't seem like we have a workable alternative for Non devs anyways.

There's likely a different approach we can take that would be useful to Non devs that doesn't require coming up with string version formats.

**Comment by iionly** [comment]
A suggestion I made (but that seem to have gone lost within my postings) would be:

"2.0 or newer Elgg 2.X"

I don't know if using "2.X" as reference to any version of Elgg 2 newer than 2.0 but only to version of Elgg 2 (and not 3.X etc.) would be easier to understand. Also, the string might be too long. But I think with having users of plugins (and NOT developers) in mind it might be clearer than "^2.0".

I can't follow the idea of "pulling the value from composer.json". First, this would require all plugins having a composer.json and, secondly, any information about the required Elgg version being included in composer.json. Finally, I hadn't even the process of adding a new plugin / plugin release in mind but rather the information shows on plugin pages which would still require to select a specific string and not rely on possibly different values from composer.json.

I'm beginning to regret to have opened this issue. Devs, non-devs, prove it... not in the mood for this kind of discussion.

**Comment by ewinslow** [comment]
@iionly yea, that could work. It's a bit wordy, but clear. Some other options:

 * 2.1+
 * >= 2.1
 * >= 2.1, < 3.0
 * 2.1-2.x
 * 2.1 or greater

I don't like the `>=` symbols at all... too mathy.

2.1 - 2.x seems sufficiently clear that the plugin is not expected to work on 3.x without being wordy.

You could also imagine plugins that work across major versions doing something like "2.1 - 3.x".

WDYT?

**Comment by ewinslow** [comment]
So you could imagine "Tidypics for Elgg 2.1 - 2.x"

**Comment by iionly** [comment]
"Tidypics for Elgg 2.1 - 2.x" would be the best of the suggestions (any logical operators are confusing and the next major version must be excluded).

But I'm sorry to say but I'm no longer convinced that it's a good idea trying to express that a plugin release is compatible to more than a single release. At least, it won't work with a single expression for all occasions within the community_plugin code. As I see it the code isn't ready to handle it and there would be changes to be made first to handle it correctly.

**Comment by ewinslow** [comment]
Yea, certainly would require some updates to community code...

**Comment by iionly** [comment]
Presently not intested that much anymore on this issue - mainly lacking the time to looking into what other changes might be necessary. So, closing for now.
